### PR TITLE
[generic_config_updater]: Fix emptying a leaf-list writing an invalid value to CONFIG_DB

### LINKED ISSUE #27818: Bug:GCU: emptying a leaf-list writes invalid "" to CONFIG_DB instead of removing the field
### Is it platform specific

generic

### Importance or Severity

Critical

### Description of the bug

When a Generic Config Updater (`config apply-patch`) JSON patch reduces a **leaf-list to empty**
(e.g. removing its last element), the change is committed to CONFIG_DB as an **empty string**
`<field>@ = ""` instead of the field being removed.

The patch sorter converts "remove the last leaf-list item" into a single
`{"op":"replace","path":".../<leaf-list>","value":[]}` move (introduced by
sonic-utilities PR https://github.com/sonic-net/sonic-utilities/pull/4478,
`BulkLeafListMoveGenerator`). `ConfigDBConnector` then serializes the empty list `[]` to an empty
string, so CONFIG_DB ends up holding `<field>@ = ""`.

That empty `@` field is an invalid representation. The correct CONFIG_DB form of an empty
leaf-list is the **field being absent**. When the value is read back,
`"".split(",") == ['']` — it becomes a single empty-string element that fails YANG validation.

Consequences:
1. The config becomes un-loadable: every subsequent `config apply-patch` on that table fails YANG
   validation with `"" does not conform`.
2. Daemons consuming the table log errors (e.g. `bgpcfgd: invalid input[prefixes_v4]:''`).

Affected component: **sonic-utilities** `generic_config_updater/patch_sorter.py`
(`BulkLeafListMoveGenerator`), introduced by #4478. `BGP_ALLOWED_PREFIXES` is used below only as a
convenient example of a leaf-list table; the defect is generic to any leaf-list.

Suggested fix: in `BulkLeafListMoveGenerator`, when the target leaf-list is empty, emit a `REMOVE`
of the field (so it becomes absent) instead of `REPLACE = []`.

### Steps to Reproduce

Single-ASIC SONiC DUT (image with #4478). Run as `admin`:

```bash
# 0. clean slate
sonic-db-cli CONFIG_DB del "BGP_ALLOWED_PREFIXES|DEPLOYMENT_ID|0"

# 1. seed a leaf-list with ONE item
cat > /tmp/add.json <<'EOF'
[{"op":"add","path":"/BGP_ALLOWED_PREFIXES","value":{
   "DEPLOYMENT_ID|0":{"prefixes_v4":["10.20.0.0/16"],"prefixes_v6":["fc01:20::/64"]}}}]
EOF
sudo config apply-patch /tmp/add.json
sonic-db-cli CONFIG_DB hgetall "BGP_ALLOWED_PREFIXES|DEPLOYMENT_ID|0"
#   -> {'prefixes_v4@': '10.20.0.0/16', 'prefixes_v6@': 'fc01:20::/64'}

# 2. remove the LAST v4 prefix  -> empties the leaf-list
cat > /tmp/empty.json <<'EOF'
[{"op":"remove","path":"/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4/0"}]
EOF
sudo config apply-patch /tmp/empty.json

# 3. BUG: field is stored as "" instead of being removed
sonic-db-cli CONFIG_DB hgetall "BGP_ALLOWED_PREFIXES|DEPLOYMENT_ID|0"
#   -> {'prefixes_v4@': '', 'prefixes_v6@': 'fc01:20::/64'}   <-- prefixes_v4@ should be ABSENT

# 4. config is now un-loadable: any further GCU op fails on the ""
cat > /tmp/next.json <<'EOF'
[{"op":"replace","path":"/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4","value":["10.40.0.0/16"]}]
EOF
sudo config apply-patch /tmp/next.json
#   -> fails: Unsatisfied pattern - "" does not conform ... /prefixes_v4

# cleanup
sonic-db-cli CONFIG_DB del "BGP_ALLOWED_PREFIXES|DEPLOYMENT_ID|0"
```
The two confirming checks (no BGP feature dependency) are Step 3 (prefixes_v4@ present but
blank) and Step 4 (follow-up patch fails with "" does not conform). The bgpcfgd log line in
Step 2 is an additional symptom that appears when the allow-list feature is enabled.

### Actual Behavior and Expected Behavior

**Actual:** Removing the last element of a leaf-list leaves `prefixes_v4@ = ""` in CONFIG_DB (an
invalid empty-string list, field still present). The config can no longer be loaded — subsequent
`config apply-patch` calls fail with `"" does not conform`, and bgpcfgd logs
`invalid input[prefixes_v4]:''`.

{'prefixes_v4@': '', 'prefixes_v6@': 'fc01:20::/64'}

**Expected:** Emptying a leaf-list should **remove the field** (`prefixes_v4@` absent), the valid
CONFIG_DB representation of "no items". Subsequent operations should succeed and no error should be
logged.

{'prefixes_v6@': 'fc01:20::/64'}      # prefixes_v4@ removed entirely

### Relevant log output

```shell
# GCU converts the leaf-list emptying into a replace-with-empty-list (PR #4478):
NOTICE config: Patch Applier: The localhost patch was converted into 1 change:
NOTICE config: Patch Applier: localhost: applying 1 change in order:
NOTICE config: Patch Applier:   * [{"op": "replace", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4", "value": []}]

# bgpcfgd receives the empty value and rejects it:
ERR bgp#bgpcfgd: BGPAllowListMgr::Received BGP ALLOWED 'SET' message with invalid input[prefixes_v4]:''

# any subsequent GCU op cannot load the now-invalid config:
apply-patch failed for localhost: Data Loading Failed
failed to parse data tree: Unsatisfied pattern - "" does not conform to "(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])/(([0-9])|([1-2][0-9])|(3[0-2]))( (le|ge) (([0-9])|([1-2][0-9])|(3[0-2])))?".: Data path: /sonic-bgp-allowed-prefix:sonic-bgp-allowed-prefix/BGP_ALLOWED_PREFIXES/BGP_ALLOWED_PREFIXES_LIST[deployment='DEPLOYMENT_ID'][id='0']/prefixes_v4 (line 1)
```

### Output of `show version`, `show techsupport`

```shell

```

### Attach files (if any)

_No response_

### PR DESCRIPTION
**Pull request by rimunagala**
#### What I did

Fixes [sonic-net/sonic-buildimage#27818](https://github.com/sonic-net/sonic-buildimage/issues/27818).

Emptying a leaf-list through `config apply-patch` left an invalid value in CONFIG_DB and made the patch application fail. Removing the last item of `BGP_ALLOWED_PREFIXES|DEPLOYMENT_ID|0` `prefixes_v4` wrote `prefixes_v4@ = ""`, and the command reported `after applying patch to config, there are still some parts not updated`. Every later patch touching that field then failed YANG validation with `"" does not conform`.

#### How I did it

CONFIG_DB has no representation for an empty leaf-list. `set_entry` serializes `[]` to an empty string, which is read back as `[""]`, so a target config that asks for `[]` can never be satisfied and the final `verify_same_json` in `PatchApplier.apply` always mismatches. The canonical way to express "this leaf-list has no items" is for the field to be absent.

The target config is always derived through `PatchWrapper`, so a ConfigDB specific simulation entry point was added there, and empty leaf-lists are dropped in it:

```python
def simulate_config_db_patch(self, patch, config_db):
    return remove_empty_leaf_lists(self.simulate_patch(patch, config_db))
```

This is used everywhere the result is meant to be a ConfigDB target: `PatchApplier.apply`, all three sorters (`StrictPatchSorter.sort`, `NonStrictPatchSorter.sort`, `PatchSorter.sort`) and `convert_config_db_patch_to_sonic_yang_patch`. The sorter then produces a field level `remove` instead of a `replace` with an empty list, the change applier writes the entry without the field, and the readback matches the target.

`simulate_patch` itself stays shape agnostic, so the `--format SONICYANG` path is unaffected. In SonicYang json a list holds yang list entries rather than leaf-list items, so ConfigDB canonicalization must not be applied there.

The guard in `BulkLeafListMoveGenerator` is kept as defence in depth, so that generator never emits a bulk `replace` with an empty list.

> Note on scope: this only reproduces on images whose `sonic_yang` uses the libyang3 python bindings (`import libyang as ly`). That holds for master and for 202605, so both are affected. On branches still on libyang 1.0.73 (`import yang as ly`) the empty list is rejected during validation before the sorter runs, so the bad value never reaches CONFIG_DB.

#### How to verify it

Verified end to end on `docker-sonic-vs` built from master (`db028339e`) using the reproduction steps from the issue.

Before, step 2 emits a `replace` with an empty list, `apply-patch` fails, and CONFIG_DB is left corrupted:

```
=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
apply-patch failed for localhost: localhost: after applying patch to config, there are still some parts not updated
Error: Failed to apply patch on the following scopes:
Patch Applier: The localhost patch was converted into 1 change:
Patch Applier:   * [{"op": "replace", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4", "value": []}]

=== STEP 3: CONFIG_DB state ===
{'prefixes_v4@': '', 'prefixes_v6@': 'fc01:20::/64'}
```

After, step 2 emits a field level `remove`, the patch applies, and the field is gone:

```
=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
Patch Applier: The localhost patch was converted into 1 change:
Patch Applier:   * [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4"}]
Patch applied successfully.

=== STEP 3: CONFIG_DB state ===
{'prefixes_v6@': 'fc01:20::/64'}
```

Re-adding the leaf-list afterwards works, so the config is fully recoverable:

```
=== re-add via add ===
Patch applied successfully.
{'prefixes_v6@': 'fc01:20::/64', 'prefixes_v4@': '10.40.0.0/16'}
```

Edge case where the leaf-list is the only field in the entry also applies cleanly, the entry is kept as `NULL: NULL`, which is the normal ConfigDB representation of an entry with no fields.

Unit tests, run in the same container against `tests/generic_config_updater/`:

| | result |
| --- | --- |
| master baseline | 554 passed, 81 subtests passed |
| this PR | 558 passed, 81 subtests passed |

Five tests are added and one pre-existing test is updated, so the net delta is four.

The updated one is `test_generate__leaf_list_all_items_removed__single_replace_move` from #4478. It asserted that emptying a leaf-list produces a single bulk `replace` with an empty list, which is exactly the value CONFIG_DB cannot store, so that assertion has to be reversed. It is renamed to `test_generate__leaf_list_all_items_removed__no_bulk_replace_move` and now expects no bulk move, leaving the transition to the granular removal generators. The batching optimisation from #4478 is untouched for every non-empty leaf-list, which is still covered by `test_generate__leaf_list_items_removed__single_replace_move` and `test_generate__leaf_list_items_added__single_replace_move`.

#### Which release branch to backport

- [x] 202605

Updated following @dgsudharsan's request. 202605 ships the libyang3 based `sonic_yang`, so it is affected and the fix is needed there. Reproduction and fix are both confirmed on a real 202605 image, see **Test result**.

The cherry-pick applies without conflict, but #4423 (`33c1c482`) should be picked first. It is a one line change to `_traverse_current_list` that 202605 never received, and without it the test added here fails. Details in **Test result**.

#### Tested branch (Please provide the tested image version)

- [x] master, `docker-sonic-vs` at `db028339e`
- [x] 202605, `docker-sonic-vs` at `0b2faef94`

#### Test result

master: PASS. `docker-sonic-vs` at `db028339e`. Issue reproduction end to end plus `tests/generic_config_updater/` at 558 passed against a 554 passed baseline, as described in **How to verify it**.

202605: PASS, with #4423 as a prerequisite.

Image is `docker-sonic-vs` built from 202605, reporting `branch: '202605'`, `commit_id: '0b2faef94'`. That image has `BulkLeafListMoveGenerator` from #4478 and the libyang3 binding (`sonic_yang.py:3:import libyang as ly`) and does not have this fix, so it is exposed.

Stock 202605, bug present:

```
=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
apply-patch failed for localhost: localhost: after applying patch to config, there are still some parts not updated
Patch Applier: The localhost patch was converted into 1 change:
Patch Applier:   * [{"op": "replace", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4", "value": []}]

=== STEP 3: CONFIG_DB state ===
{'prefixes_v4@': '', 'prefixes_v6@': 'fc01:20::/64'}
```

202605 with this PR cherry-picked, bug fixed:

```
=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
Patch Applier: The localhost patch was converted into 2 changes:
Patch Applier:   * [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4/0"}]
Patch Applier:   * [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4"}]
Patch applied successfully.

=== STEP 3: CONFIG_DB state ===
{'prefixes_v6@': 'fc01:20::/64'}
```

CONFIG_DB ends correct, but the sorter takes two changes instead of one, because 202605 is missing #4423 which relaxed `_traverse_current_list` from `len(ptr) == 0` to `len(ptr) <= 1`. That difference is visible to the test added here, `test_patch_sorter__remove_last_bgp_allowed_prefix__removes_field_instead_of_emptying_it`, which asserts the single change form.

`tests/generic_config_updater/` run in the same 202605 container:

| 202605 tree | result |
| --- | --- |
| baseline, nothing picked | 549 passed, 81 subtests passed |
| this PR only | 552 passed, 81 subtests passed, 1 failed |
| #4423 + this PR | 553 passed, 81 subtests passed, 0 failed |

With #4423 picked first, 202605 matches master exactly, a single field level `remove`:

```
=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
Patch Applier: The localhost patch was converted into 1 change:
Patch Applier:   * [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4"}]
Patch applied successfully.

=== STEP 3: CONFIG_DB state ===
{'prefixes_v6@': 'fc01:20::/64'}
```

#4423 is already included in 202511 and has a cherry-pick PR created for msft-202601, but it was never taken to 202605, so picking it there is a gap fix in its own right. Both cherry-picks apply without conflict.

#### Description for the changelog

Fix `config apply-patch` leaving an invalid empty value in CONFIG_DB when a patch empties a leaf-list.

### CODE REVIEW

**Comment by rimunagala** [comment]
@vaibhavhd thanks for asking for the repro results, running them changed the fix.

I brought up `docker-sonic-vs` built from master (`db028339e`) and ran the exact steps from sonic-net/sonic-buildimage#27818. The first version of this PR only guarded `BulkLeafListMoveGenerator`, and on a real switch that turned out **not** to fix the bug. It changed which operation the sorter emitted, but CONFIG_DB was still left with `prefixes_v4@ = ''` and `apply-patch` still failed. So I re-rooted the fix. Details below.

### Root cause

CONFIG_DB has no representation for an empty leaf-list. In `change_applier.set_config`, `config_db.set_entry()` serializes `[]` to an empty string, and `get_config_db_as_json()` reads it back as `['']`. So a target config that asks for `[]` can never be satisfied, and the final `verify_same_json(target_config, new_config)` in `PatchApplier.apply` compares `[]` against `['']` and always mismatches. That is the `after applying patch to config, there are still some parts not updated` error.

This is why a sorter-only guard cannot work: whichever operation the sorter picks, the resulting config dict still holds `prefixes_v4: []`.

The canonical way to say "this leaf-list has no items" is for the field to be absent. `PatchWrapper.simulate_patch` is the single place the target config is derived, and it is shared by `PatchApplier.apply` and all three `PatchSorter` call sites, so empty leaf-lists are now dropped there. The sorter then emits a field level `remove` and everything downstream lines up.

### Repro results

Scope note first: this only reproduces on images whose `sonic_yang` uses the libyang3 bindings (`import libyang as ly`), which is master. On branches still on libyang 1.0.73 (`import yang as ly`) the empty list is rejected during validation before the sorter runs, so the bad value never reaches CONFIG_DB. I confirmed this against a `202511` image, which cannot reproduce.

**Before, stock master:**

```
=== STEP 1: seed leaf-list with one item ===
Patch applied successfully.
{'prefixes_v4@': '10.20.0.0/16', 'prefixes_v6@': 'fc01:20::/64'}

=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
apply-patch failed for localhost: localhost: after applying patch to config, there are still some parts not updated
Error: Failed to apply patch on the following scopes:
Patch Applier: localhost: Patch: [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4/0"}]
Patch Applier: The localhost patch was converted into 1 change:
Patch Applier: localhost: applying 1 change in order:
Patch Applier:   * [{"op": "replace", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4", "value": []}]

=== STEP 3: CONFIG_DB state ===
{'prefixes_v4@': '', 'prefixes_v6@': 'fc01:20::/64'}
```

`prefixes_v4@` is the empty string, which matches the issue report.

**After, with this PR:**

```
=== STEP 1: seed leaf-list with one item ===
Patch applied successfully.
{'prefixes_v4@': '10.20.0.0/16', 'prefixes_v6@': 'fc01:20::/64'}

=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
Patch Applier: localhost: Patch: [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4/0"}]
Patch Applier: The localhost patch was converted into 1 change:
Patch Applier: localhost: applying 1 change in order:
Patch Applier:   * [{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4"}]
Patch applied successfully.

=== STEP 3: CONFIG_DB state ===
{'prefixes_v6@': 'fc01:20::/64'}
```

The field is removed instead of being set to an empty string, and the patch applies successfully.

**Recovery, re-adding the leaf-list afterwards:**

```
=== re-add via add ===
Patch applied successfully.
{'prefixes_v6@': 'fc01:20::/64', 'prefixes_v4@': '10.40.0.0/16'}
```

One thing worth calling out. After the fix the field is genuinely absent, so a follow up `{"op": "replace", "path": ".../prefixes_v4", ...}` is correctly rejected with `can't replace a non-existent object 'prefixes_v4'`. That is RFC 6902 behaviour for `replace` on a missing path, not a regression, and `add` is the right operation there.

**Edge case,** leaf-list is the only field in the entry. The patch applies and the entry is kept as `{'NULL': 'NULL'}`, which is the normal ConfigDB representation of an entry with no fields.

### Unit tests

Run in the same container against `tests/generic_config_updater/`:

| | result |
| --- | --- |
| master baseline | 554 passed, 81 subtests passed |
| this PR | 557 passed, 81 subtests passed |

The delta is exactly the three tests added here, no existing test changed behaviour.

I have updated the PR title, description and commits to reflect the real fix. Happy to add anything else you would like to see.

**Comment by rimunagala** [comment]
Thanks for the review, the naming point was a good catch and chasing it turned up something a bit more than a docstring issue.

`PatchWrapper.simulate_patch` is not only used for ConfigDB. `convert_sonic_yang_patch_to_config_db_patch` calls it with **SonicYang** shaped json:

```python
def convert_sonic_yang_patch_to_config_db_patch(self, patch):
    current_yang = self.config_wrapper.get_sonic_yang_as_json()
    target_yang = self.simulate_patch(patch, current_yang)
```

That is a live path, it is what `config apply-patch --format SONICYANG` goes through via `SonicYangDecorator`. The two shapes are different:

```
ConfigDB    table -> key -> field           a list valued field here is a leaf-list
SonicYang   module -> container -> LIST     a list here is a yang list of entries,
                                            the real leaf-lists sit one level deeper
```

So the previous revision would have dropped a `VLAN_LIST` that a patch had emptied, which is not a leaf-list, while missing the actual leaf-lists inside the list entries. Almost certainly harmless in practice, but it is an unvalidated behaviour change on a path unrelated to this bug.

I have pushed a revision that scopes the normalization instead of renaming it:

```python
def simulate_patch(self, patch, jsonconfig):
    return patch.apply(jsonconfig)

def simulate_config_db_patch(self, patch, config_db):
    return remove_empty_leaf_lists(self.simulate_patch(patch, config_db))
```

`simulate_config_db_patch` is now used at every site whose result is a ConfigDB target, `PatchApplier.apply`, `PatchSorter.sort`, `NonStrictPatchSorter.sort` and `convert_config_db_patch_to_sonic_yang_patch`. The SonicYang call site keeps plain `simulate_patch`.

On the naming, once the helper is only reachable with ConfigDB shaped json the name is accurate, because at `table -> key -> field` a list valued field **is** a leaf-list. So rather than renaming to something neutral I made the precondition explicit in the docstring:

```
Expects ConfigDB shaped json, i.e. table -> key -> field. At that shape a
list valued field is always a leaf-list, so every empty list found here is
an empty leaf-list. Do not call this with SonicYang shaped json, where a
list is a yang list of entries rather than a leaf-list.
```

Renaming alone would have documented the ambiguity, scoping removes it.

Added a test that locks the boundary in, `simulate_patch` on SonicYang json leaves an emptied `VLAN_LIST` untouched, and moved the two normalization tests onto `simulate_config_db_patch`.

Re-verified after the change: the end to end reproduction on `docker-sonic-vs` from master still ends with `prefixes_v4@` absent and `Patch applied successfully`, and `tests/generic_config_updater/` is at 558 passed against the 554 passed master baseline, the delta being exactly the four tests added here.

On the merge assessment, all CI checks were passing on the previous revision, including Azure Build Python3, diff coverage, static analysis, DCO and EasyCLA. I will confirm the same once this revision finishes.

One more data point on the "already corrupted switch" question, which is worth recording: in the before run, step 4 patched successfully over `prefixes_v4@ = ''`, so a switch that already hit this bug recovers with a normal patch, no manual CONFIG_DB surgery needed.

File: generic_config_updater/patch_sorter.py [hunk 1]
```diff
@@ -1354,7 +1354,12 @@ def _traverse(self, diff, current_ptr, target_ptr, tokens):
...
             if isinstance(current_val, list) and isinstance(target_val, list):
                 # Only handle leaf-lists (lists of scalars)
+                # An empty target leaf-list is deliberately excluded: a bulk
+                # REPLACE with [] serializes to an empty string in CONFIG_DB,
+                # which is an invalid state. Those transitions are left to the
+                # granular removal generators, which remove the field instead.
                 if (current_val != target_val and
+                        len(target_val) > 0 and
                         self._is_leaf_list(current_val) and
                         self._is_leaf_list(target_val)):
```
[the remark this answers is not in the document]
Comment (rimunagala) [reply generic_config_updater/patch_sorter.py:1364]: Good catch to raise this, and it was the right thing to worry about for the earlier revision of this PR. It is closed now, for two independent reasons.

The main one is that the fix moved from the generator to the target itself. `PatchApplier.apply` and all three sorters now derive their target through `PatchWrapper.simulate_config_db_patch`, which drops empty leaf-lists, so by the time any move generator or extender runs, `diff.target_config` no longer contains `prefixes_v4: []` at all. There is no empty list left anywhere for an extender to lift into a `REPLACE`.

The second is that `UpperLevelMoveExtender` could not produce it even if one were present. `_get_upper_operation` walks the tokens against both configs and returns as soon as a token is missing from the target:

```python
for token in tokens:
    if token not in current_ptr:
        return OperationType.ADD
    current_ptr = current_ptr[token]
    if token not in target_ptr:
        return OperationType.REMOVE
    target_ptr = target_ptr[token]

return OperationType.REPLACE
```

With the normalized target the leaf-list field is absent, so extending a per-item remove up to the field level yields `REMOVE` on the field, not `REPLACE` with `[]`.

The DFS fallback path you describe is covered end to end by `test_patch_sorter__remove_last_bgp_allowed_prefix__removes_field_instead_of_emptying_it`, which drives the full `PatchSorter.sort` for the last item of a leaf-list, asserts the single emitted change is

```json
{"op": "remove", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4"}
```

then replays every change and validates the intermediate config against YANG, and finally asserts the replayed result equals the simulated target. Any move, extended or not, that reintroduced an empty list would fail that last assertion.

The `len(target_val) > 0` guard in `BulkLeafListMoveGenerator` is kept as defence in depth rather than as the fix.

**Comment by rimunagala** [comment]
Housekeeping update, no functional change to the fix.

The branch was squashed into a single commit on top of current master (`666f192d`). The tree is byte identical to the previously reviewed `8baf7ba3`, the two earlier commits were intermediate steps of an approach this PR no longer takes and were only noise for the log.

Two review-facing things I want to call out rather than leave buried in the diff:

@rookie-who this PR reverses one assertion from #4478. `test_generate__leaf_list_all_items_removed__single_replace_move` expected `BulkLeafListMoveGenerator` to emit a single bulk `replace` with an empty list when a leaf-list is emptied, and that empty list is precisely the value CONFIG_DB cannot store, so the expectation had to flip. It is renamed to `test_generate__leaf_list_all_items_removed__no_bulk_replace_move` and now expects no bulk move, leaving that transition to the granular removal generators. The batching optimisation itself is untouched for every non-empty leaf-list and is still covered by `test_generate__leaf_list_items_removed__single_replace_move` and `test_generate__leaf_list_items_added__single_replace_move`. Would appreciate your eyes on that specifically.

The open Copilot thread about `UpperLevelMoveExtender` synthesizing a parent level `REPLACE ... []` is answered inline and resolved. Short version: the fix now normalizes the ConfigDB target before sorting, so `diff.target_config` never contains an empty list for an extender to lift, and `_get_upper_operation` returns `REMOVE` for a token absent from the target anyway.

Re-verified at the squashed commit: `generic_config_updater` suite 558 passed against a 554 passed baseline, flake8 clean on the diff, and the issue reproduction re-run end to end on `docker-sonic-vs` from master with the container code md5 checked against the commit.

**Review by vaibhavhd — APPROVED** [review]

**Comment by dgsudharsan** [comment]
@rimunagala Please test it on 202605 as this fix is needed for there as well. I added cherry-pick tag

**Comment by rimunagala** [comment]
@dgsudharsan tested on 202605 as you asked. Full evidence is now in the description under **Test result**, and the **Tested branch** section is ticked so the build bot should be satisfied. Summary, plus one thing the cherry-pick needs.

202605 is genuinely affected, and my earlier "master only" note in the description was wrong. I have corrected it. The image I used reports `branch: '202605'`, `commit_id: '0b2faef94'`, and it carries both `BulkLeafListMoveGenerator` from #4478 and the libyang3 based `sonic_yang` (`sonic_yang.py:3:import libyang as ly`), which is the combination that triggers this. Stock 202605 reproduces the issue exactly:

```
=== STEP 2: remove the LAST prefixes_v4 item (empties the leaf-list) ===
apply-patch failed for localhost: localhost: after applying patch to config, there are still some parts not updated
Patch Applier:   * [{"op": "replace", "path": "/BGP_ALLOWED_PREFIXES/DEPLOYMENT_ID|0/prefixes_v4", "value": []}]

=== STEP 3: CONFIG_DB state ===
{'prefixes_v4@': '', 'prefixes_v6@': 'fc01:20::/64'}
```

With this PR cherry-picked, the patch applies and the field is gone, so the fix does what it should on 202605.

One caveat worth flagging before the cherry-pick goes in: **#4423 (`33c1c482`) should be picked to 202605 first.** It is a one line change to `_traverse_current_list`, `len(ptr) == 0` to `len(ptr) <= 1`. 202605 never received it, although it is already included in 202511 and has a cherry-pick PR for msft-202601. Without it this fix is still functionally correct, CONFIG_DB ends identical, but the sorter emits two changes instead of one, and the test this PR adds, `test_patch_sorter__remove_last_bgp_allowed_prefix__removes_field_instead_of_emptying_it`, asserts the single change form and fails.

`tests/generic_config_updater/` run in the same 202605 container:

| 202605 tree | result |
| --- | --- |
| baseline, nothing picked | 549 passed |
| this PR only | 552 passed, 1 failed |
| #4423 + this PR | 553 passed, 0 failed |

With #4423 first, 202605 behaves exactly like master, a single field level `remove`, and the suite is green. Both cherry-picks apply without conflict. I have added the `Request for 202605 Branch` label to #4423 so it can follow the same flow. Worth noting the ordering: #4423 needs to be in 202605 before this one, otherwise the cherry-pick of this PR will fail its own unit test in 202605 CI.

One related item so it does not surprise anyone: sonic-net/sonic-mgmt#26865 needs to land on 202605 as well. 202605's `bgp_prefix_tc1_xfail` entry is byte identical to master's, and #26494 already removed the xfail there, so `test_bgp_prefix_tc1_suite` will start failing on 202605 the moment this fix arrives. I have ticked 202605 on that PR.

**Comment by rookie-who** [comment]
**Approved for 202605 cherry-pick.**

Must-have bug fix: GCU fails to empty leaf-lists, writing invalid values to CONFIG_DB breaks config updates. PR has been tested on the 202605 branch and has code review approval on master.

*Review by @vaibhavhd via automated tooling*
